Removed MinHops parameter from gossip.

spencerkimball opened the pull request:
It was originally intended to be a part of the network tightening
mechanism but ended up not being used and was getting in the way
of correctly determining which info was fresh between nodes.

The previous broken use of MinHops was preventing efficient relay
of gossip info, causing network convergence to lag.

Fixes #4500

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.svg" height="40" alt="Review on Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/4502)

<!-- Reviewable:end -->

spencerkimball committed:
Removed MinHops parameter from gossip.

It was originally intended to be a part of the network tightening
mechanism but ended up not being used and was getting in the way
of correctly determining which info was fresh between nodes.

The previous broken use of MinHops was preventing efficient relay
of gossip info, causing network convergence to lag.

spencerkimball commented:
@tamird weirdly when I run `gossip/simulation/gossip.go --size=medium` with this change, the grpc connections fail to close properly and it hangs when it's finished. I'm not sure why that would be the case, except perhaps this change makes things happen faster and is exercising some sort of race condition.

tamird commented:
That does seem weird. I'll look into it first thing tomorrow.

---

Reviewed 9 of 9 files at r1.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[gossip/gossip.proto, line 67 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4502#-KAr2sI7QHafVDylgunO:-KAr2sI85g8a3Jd5o1ry:124549485) ([raw file](https://github.com/cockroachdb/cockroach/blob/74c8d84262c849ea87150c48244301ea390138d5/gossip/gossip.proto#L67)):*
While you're here:
`map<int32,int64> highWaterStamps = 6 [(gogoproto.castkey) = "github.com/cockroachdb/cockroach/roachpb.NodeID", (gogoproto.nullable) = false];`
or leave a todo

---

*[gossip/info.go, line 27 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4502#-KAr1y1D8M-rJWt5Vlb2:-KAr1y1D8M-rJWt5Vlb3:1370864376) ([raw file](https://github.com/cockroachdb/cockroach/blob/74c8d84262c849ea87150c48244301ea390138d5/gossip/info.go#L27)):*
`return i.Origstamp > highWaterStamp`

---

*[gossip/infostore.go, line 191 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4502#-KAr4DTs4rsv1kNCfV7Y:-KAr4DTtfWy2V1WPTA1-:2094011231) ([raw file](https://github.com/cockroachdb/cockroach/blob/74c8d84262c849ea87150c48244301ea390138d5/gossip/infostore.go#L191)):*
no need to check `ok` here

---

*[gossip/simulation/network.go, line 125 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4502#-KAr3bcBkEzPPIlfIxkT:-KAr3bcC8QZKZE1kAV7K:-1321265596) ([raw file](https://github.com/cockroachdb/cockroach/blob/6135537594bdf3936806acac05eab0a6c1f5080d/gossip/simulation/network.go#L125)):*
why make these public?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4502#-:-KAr4c2WMWldsrHGpmve:-894045113)*
<!-- Sent from Reviewable.io -->

tamird commented:
I see a bunch of error spew at shutdown, but no hang. The error spew is sadly expected, as grpc doesn't expose a synchronous `Close` on the connections (the assumption is that these connections are forever). So we call `Close` but the connection may already be in the reconnect loop, which can produce the log spew when the remote has shut down.

Can you get a stack trace from the hang you see?

---

Review status: all files reviewed at latest revision, 4 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4502#-:-KAtbjadhUaKnjrU_Din:-436314110)*
<!-- Sent from Reviewable.io -->

spencerkimball commented:
Can't get it to fail now...

---

Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[gossip/simulation/network.go, line 125 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4502#-KAr3bcBkEzPPIlfIxkT:-KAvo-VCTsZzcVfdSMsH:1441874953) ([raw file](https://github.com/cockroachdb/cockroach/blob/6135537594bdf3936806acac05eab0a6c1f5080d/gossip/simulation/network.go#L125)):*
They're used from main package.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4502#-:-KAvo1yqwNj8ZNMv-wBu:-1966336232)*
<!-- Sent from Reviewable.io -->

spencerkimball committed:
Change convergence test to verify just that network converges

We now warn if it doesn't converge as quickly as expected but
it's proven extremely difficult to orchestrate the simulation
such that we can guarantee that with arbitrarily slow test envs.

Fixes #4500

tamird commented:
LGTM. Add "Fixes #4500" to the second commit message.

---

Reviewed 9 of 9 files at r2.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[gossip/convergence_test.go, line 29 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4502#-KAvsTEI6DOkpclIg6KV:-KAvsTEI6DOkpclIg6KW:1260114251) ([raw file](https://github.com/cockroachdb/cockroach/blob/ec268c198c92b74429d0a58859fd0fbb33a5e3a1/gossip/convergence_test.go#L29)):*
s/ t / _ / since it's unused

---

*[gossip/gossip.go, line 939 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4502#-KAvrZZKMJmNEyf3evQx:-KAvrZZKMJmNEyf3evQy:-1542324587) ([raw file](https://github.com/cockroachdb/cockroach/blob/ec268c198c92b74429d0a58859fd0fbb33a5e3a1/gossip/gossip.go#L939)):*
remove this

---

*[gossip/simulation/network.go, line 125 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4502#-KAr3bcBkEzPPIlfIxkT:-KAvqvDM3IzKM5qPr5RA:-1812980965) ([raw file](https://github.com/cockroachdb/cockroach/blob/6135537594bdf3936806acac05eab0a6c1f5080d/gossip/simulation/network.go#L125)):*
Huh, we have `time go build ./gossip/simulation/...` in `circle-deps.sh` but that must not build `main`.  We need to add `time go build ./gossip/simulation/gossip.go` apparently.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4502#-:-KAvsOBNq8AQlbnvj3at:-443052757)*
<!-- Sent from Reviewable.io -->

bdarnell commented on gossip/convergence_test.go:
```diff
@@ -45,6 +46,5 @@ func verifyConvergence(numNodes, maxCycles int, t *testing.T) {
 func TestConvergence(t *testing.T) {
 	defer leaktest.AfterTest(t)
-	t.Skipf("TODO(pmattis): #4500")
 	verifyConvergence(10, 100, t)
```
How many iterations are now required? Can we lower the "100" here? (it was 20 until a few weeks ago)
